[css-text-4] Update `text-wrap` properties to latest spec

nt1m opened the pull request:
#### 3322e6b61d68491df248d444fb09f3c990ad0414
<pre>
[css-text-4] Update `text-wrap` properties to latest spec
<a href="https://bugs.webkit.org/show_bug.cgi?id=263432">https://bugs.webkit.org/show_bug.cgi?id=263432</a>
rdar://117248327

Reviewed by Darin Adler.

There is now a new arrangement for the longhands that this commit implements:
- white-space is a shorthand of white-space-collapse &amp; text-wrap-mode
- text-wrap is a shorthand of text-wrap-mode &amp; text-wrap-style

The previous text-wrap property values are split between its longhands:
- text-wrap-style takes the values auto / balance / pretty / stable. This property is disabled behind a preference.
- text-wrap-mode takes the values wrap / nowrap.

This commit also enables by default the longhands for white-space now that layout support for all the different combinations is implemented.
It also removes the related workarounds.

WPTs are updated to use shortest serialization.

Specification: <a href="https://drafts.csswg.org/css-text-4/">https://drafts.csswg.org/css-text-4/</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/text-wrap-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/text-wrap-valid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/white-space-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/white-space-shorthand-text-wrap-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/white-space-shorthand-text-wrap.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/white-space-shorthand.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/white-space-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/white-space.html:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::toCSSValueID):
(WebCore::fromCSSValueID):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::textWrapShorthandValue const):
(WebCore::ComputedStyleExtractor::whiteSpaceShorthandValue const):
(WebCore::ComputedStyleExtractor::propertyValue const):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/css/ComputedStyleExtractor.h:
* Source/WebCore/css/ShorthandSerializer.cpp:
(WebCore::ShorthandSerializer::serialize):
(WebCore::ShorthandSerializer::serializeTextWrap const):
(WebCore::ShorthandSerializer::serializeWhiteSpace const):
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::CSSParserContext::CSSParserContext):
(WebCore::add):
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::initialValueForLonghand):
(WebCore::CSSPropertyParser::consumeTextWrapShorthand):
(WebCore::CSSPropertyParser::consumeWhiteSpaceShorthand):
(WebCore::CSSPropertyParser::parseShorthand):
* Source/WebCore/css/parser/CSSPropertyParser.h:
* Source/WebCore/editing/EditingStyle.cpp:
(WebCore::EditingStyle::conflictsWithInlineStyleOfElement const):
(WebCore::EditingStyle::removeStyleFromRulesAndContext):
(WebCore::StyleChange::StyleChange):
* Source/WebCore/editing/TextManipulationController.cpp:
(WebCore::isEnclosingItemBoundaryElement):
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::createInnerTextStyle):
* Source/WebCore/html/HTMLPreElement.cpp:
(WebCore::HTMLPreElement::collectPresentationalHintsForAttribute):
* Source/WebCore/html/HTMLTableCellElement.cpp:
(WebCore::HTMLTableCellElement::collectPresentationalHintsForAttribute):
* Source/WebCore/html/HTMLTextAreaElement.cpp:
(WebCore::HTMLTextAreaElement::collectPresentationalHintsForAttribute):
* Source/WebCore/html/track/TextTrackCueGeneric.cpp:
(WebCore::TextTrackCueGenericBoxElement::applyCSSProperties):
* Source/WebCore/html/track/VTTCue.cpp:
(WebCore::VTTCueBox::applyCSSProperties):
* Source/WebCore/layout/formattingContexts/inline/InlineFormattingContext.cpp:
(WebCore::Layout::InlineFormattingContext::layout):
* Source/WebCore/layout/formattingContexts/inline/InlineQuirks.cpp:
(WebCore::Layout::InlineQuirks::trailingNonBreakingSpaceNeedsAdjustment const):
* Source/WebCore/layout/formattingContexts/inline/InlineTextItem.cpp:
(WebCore::Layout::InlineTextItem::isQuirkNonBreakingSpace const):
* Source/WebCore/layout/formattingContexts/inline/TextOnlySimpleLineBuilder.cpp:
(WebCore::Layout::TextOnlySimpleLineBuilder::isEligibleForSimplifiedTextOnlyInlineLayout):
* Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
(WebCore::Layout::TextUtil::isWrappingAllowed):
(WebCore::Layout::TextUtil::shouldTrailingWhitespaceHang):
* Source/WebCore/layout/integration/LayoutIntegrationCoverage.cpp:
(WebCore::LayoutIntegration::shouldInvalidateLineLayoutPathAfterChangeFor):
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::computeBlockPreferredLogicalWidths const):
* Source/WebCore/rendering/RenderListItem.cpp:
(WebCore::RenderListItem::computeMarkerStyle const):
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::computePreferredLogicalWidths):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::adjustStyle):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::adjustMenuListStyle const):
* Source/WebCore/rendering/line/BreakingContext.h:
(WebCore::BreakingContext::clearLineBreakIfFitsOnLine):
(WebCore::BreakingContext::initializeForCurrentObject):
(WebCore::BreakingContext::handleReplaced):
(WebCore::BreakingContext::handleText):
(WebCore::BreakingContext::trailingSpacesHang):
(WebCore::BreakingContext::canBreakAtThisPosition):
(WebCore::BreakingContext::handleEndOfLine):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::RenderStyle):
(WebCore::RenderStyle::changeRequiresLayout const):
(WebCore::RenderStyle::conservativelyCollectChangedAnimatableProperties const):
(WebCore::RenderStyle::whiteSpace const):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::textWrapMode const):
(WebCore::RenderStyle::textWrapStyle const):
(WebCore::RenderStyle::setTextWrapMode):
(WebCore::RenderStyle::setTextWrapStyle):
(WebCore::RenderStyle::textWrap const): Deleted.
(WebCore::RenderStyle::setTextWrap): Deleted.
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialTextWrapMode):
(WebCore::RenderStyle::initialTextWrapStyle):
(WebCore::RenderStyle::initialTextWrap): Deleted.
* Source/WebCore/rendering/style/TextSizeAdjustment.cpp:
(WebCore::AutosizeStatus::computeStatus):
* Source/WebCore/style/PropertyAllowlist.cpp:
(WebCore::Style::isValidMarkerStyleProperty):
(WebCore::Style::isValidCueStyleProperty):
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):
* Source/WebCore/svg/SVGTextContentElement.cpp:
(WebCore::SVGTextContentElement::collectPresentationalHintsForAttribute):

Canonical link: <a href="https://commits.webkit.org/269613@main">https://commits.webkit.org/269613@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5c6f9e8c226b4399347c6447713bfc46d6cc6412

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23056 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1138 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24170 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24971 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/21325 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23322 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/2385 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23591 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22184 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23297 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/583 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19994 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25823 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/458 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20879 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/27051 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/20059 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/20846 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21143 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/24919 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/22422 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/562 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/18347 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/29797 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/488 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/20657 "Passed tests") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/29797 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5505 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/970 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/29491 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/762 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/6049 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -3827,7 +3827,8 @@ CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap()
         new DiscretePropertyWrapper<OptionSet<TouchAction>>(CSSPropertyTouchAction, &RenderStyle::touchActions, &RenderStyle::setTouchActions),
         new DiscretePropertyWrapper<OptionSet<TextTransform>>(CSSPropertyTextTransform, &RenderStyle::textTransform, &RenderStyle::setTextTransform),
         new DiscretePropertyWrapper<WhiteSpaceCollapse>(CSSPropertyWhiteSpaceCollapse, &RenderStyle::whiteSpaceCollapse, &RenderStyle::setWhiteSpaceCollapse),
-        new DiscretePropertyWrapper<TextWrap>(CSSPropertyTextWrap, &RenderStyle::textWrap, &RenderStyle::setTextWrap),
```
I think this makes `text-wrap` no longer animatable since I don't see a change to add it to `animatableShorthandProperties`.

nt1m replied:
Good catch, fixed.

darinadler commented on Source/WebCore/css/CSSPrimitiveValueMappings.h:
```diff
@@ -1541,44 +1541,38 @@ DEFINE_TO_FROM_CSS_VALUE_ID_FUNCTIONS
 #undef FOR_EACH
 
-constexpr CSSValueID toCSSValueID(TextWrap wrap)
+constexpr CSSValueID toCSSValueID(TextWrapMode wrap)
```
Why can’t we `DEFINE_TO_FROM_CSS_VALUE_ID_FUNCTIONS` for this? Surprised it was hand-written.

nt1m replied:
It's because the enum value is NoWrap and the CSSValueID is CSSValueNowrap (notice the capitalization).

darinadler commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -577,28 +577,41 @@ static RefPtr<CSSValue> positionOffsetValue(const RenderStyle& style, CSSPropert
 
+RefPtr<CSSValue> ComputedStyleExtractor::textWrapShorthandValue(const RenderStyle& style) const
+{
+    auto textWrapMode = style.textWrapMode();
+    auto textWrapStyle = style.textWrapStyle();
```
I like taking advantage of scope to have short variable names. I would just use `mode` and `style` here, because in this context that’s enough.

nt1m replied:
I had `mode` / `style` originally, but `style` conflicts with the parameter name.

darinadler commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -577,28 +577,41 @@ static RefPtr<CSSValue> positionOffsetValue(const RenderStyle& style, CSSPropert
     return CSSPrimitiveValue::create(CSSValueAuto);
 }
 
+RefPtr<CSSValue> ComputedStyleExtractor::textWrapShorthandValue(const RenderStyle& style) const
```
I’m so annoyed that we have to write these functions. I don’t think other browsers support expressing computed style as a `CSSValue` object and all of this code ends up duplicating what the shorthand serializer does, so there are two opportunities to get it wrong. But I guess our editing code depends on this peculiar stuff?

nt1m replied:
I believe CSS Typed OM depends on this code (tested by one of the tests).

darinadler replied:
I'm surprised to hear that. Long term it's really a bummer to have to handle every shorthand twice and also to have to design a non standard CSS value structure for each shorthand when the standards describe the serialization, not objects. At some point worth looking into this to be sure it's valuable.

nt1m replied:
I'll try to see in the future if I can remove this

darinadler commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -2534,10 +2536,39 @@ bool CSSPropertyParser::consumeListStyleShorthand(bool important)
+    if (!m_range.atEnd())
+        return false;
+
+    // Fill in default values if one was missing from the multi-value syntax.
```
What would happen if instead we only added the specified properties? Would a test fail? I guess I already forgot this from when I was doing CSS property work.

nt1m replied:
We'd have no value to pass to `addProperty(` so we'd set the shorthand wouldn't do its job of resetting other longhand declarations. Tests cover this.

darinadler commented on Source/WebCore/html/HTMLInputElement.cpp:
```diff
@@ -2237,7 +2237,7 @@ RenderStyle HTMLInputElement::createInnerTextStyle(const RenderStyle& style)
 
     textBlockStyle.setWhiteSpaceCollapse(WhiteSpaceCollapse::Preserve);
-    textBlockStyle.setTextWrap(TextWrap::NoWrap);
+    textBlockStyle.setTextWrapMode(TextWrapMode::NoWrap);
```
Just confirming that we do not need to also set `text-wrap-style` here?

nt1m replied:
No, we only want to set the longhands of `white-space` which are `white-space-collapse` and `text-wrap-mode`. `text-wrap-style` should keep its default value of `auto` (there's no situation where the UA really needs to balance text by default usually).

darinadler commented on Source/WebCore/html/HTMLPreElement.cpp:
```diff
@@ -56,7 +56,7 @@ void HTMLPreElement::collectPresentationalHintsForAttribute(const QualifiedName&
     if (name == wrapAttr) {
         style.setProperty(CSSPropertyWhiteSpaceCollapse, CSSValuePreserve);
-        style.setProperty(CSSPropertyTextWrap, CSSValueWrap);
+        style.setProperty(CSSPropertyTextWrapMode, CSSValueWrap);
```
Just confirming that we do not need to also set `text-wrap-style` here?

nt1m replied:
Same here

darinadler commented on Source/WebCore/html/HTMLTableCellElement.cpp:
```diff
@@ -110,7 +110,7 @@ void HTMLTableCellElement::collectPresentationalHintsForAttribute(const Qualifie
     case AttributeNames::nowrapAttr:
         addPropertyToPresentationalHintStyle(style, CSSPropertyWhiteSpaceCollapse, CSSValueCollapse);
-        addPropertyToPresentationalHintStyle(style, CSSPropertyTextWrap, CSSValueNowrap);
+        addPropertyToPresentationalHintStyle(style, CSSPropertyTextWrapMode, CSSValueNowrap);
```
Just confirming that we do not need to also set `text-wrap-style` here?

nt1m replied:
Same here

darinadler commented on Source/WebCore/html/HTMLTextAreaElement.cpp:
```diff
@@ -135,11 +135,11 @@ void HTMLTextAreaElement::collectPresentationalHintsForAttribute(const Qualified
         if (m_wrap != NoWrap) {
             addPropertyToPresentationalHintStyle(style, CSSPropertyWhiteSpaceCollapse, CSSValuePreserve);
-            addPropertyToPresentationalHintStyle(style, CSSPropertyTextWrap, CSSValueWrap);
+            addPropertyToPresentationalHintStyle(style, CSSPropertyTextWrapMode, CSSValueWrap);
```
Just confirming that we do not need to also set `text-wrap-style` here?

nt1m replied:
Same here

darinadler commented on Source/WebCore/html/HTMLTextAreaElement.cpp:
```diff
@@ -135,11 +135,11 @@ void HTMLTextAreaElement::collectPresentationalHintsForAttribute(const Qualified
         } else {
             addPropertyToPresentationalHintStyle(style, CSSPropertyWhiteSpaceCollapse, CSSValuePreserve);
-            addPropertyToPresentationalHintStyle(style, CSSPropertyTextWrap, CSSValueNowrap);
+            addPropertyToPresentationalHintStyle(style, CSSPropertyTextWrapMode, CSSValueNowrap);
```
Just confirming that we do not need to also set `text-wrap-style` here?

nt1m replied:
Same here

darinadler commented on Source/WebCore/html/track/TextTrackCueGeneric.cpp:
```diff
@@ -153,7 +153,7 @@ void TextTrackCueGenericBoxElement::applyCSSProperties()
     setInlineStyleProperty(CSSPropertyWritingMode, cue->getCSSWritingMode(), false);
     setInlineStyleProperty(CSSPropertyWhiteSpaceCollapse, CSSValuePreserve);
-    setInlineStyleProperty(CSSPropertyTextWrap, CSSValueWrap);
+    setInlineStyleProperty(CSSPropertyTextWrapMode, CSSValueWrap);
```
Just confirming that we do not need to also set `text-wrap-style` here?

nt1m replied:
Same here

darinadler commented on Source/WebCore/html/track/VTTCue.cpp:
```diff
@@ -260,7 +260,7 @@ void VTTCueBox::applyCSSProperties()
     if (!cue->snapToLines()) {
         setInlineStyleProperty(CSSPropertyWhiteSpaceCollapse, CSSValuePreserve);
-        setInlineStyleProperty(CSSPropertyTextWrap, CSSValueNowrap);
+        setInlineStyleProperty(CSSPropertyTextWrapMode, CSSValueNowrap);
```
Just confirming that we do not need to also set `text-wrap-style` here?

nt1m replied:
Same here

darinadler commented on Source/WebCore/layout/formattingContexts/inline/InlineFormattingContext.cpp:
```diff
@@ -133,7 +133,7 @@ InlineLayoutResult InlineFormattingContext::layout(const ConstraintsForInlineCon
     };
 
-    if (root().style().textWrap() == TextWrap::Balance) {
+    if (root().style().textWrapMode() == TextWrapMode::Wrap && root().style().textWrapStyle() == TextWrapStyle::Balance) {
```
I really wish we had a boolean for `text-wrap-mode` instead of an enumeration for call sites like this.

nt1m replied:
We do have a helper method called `RenderStyle::autoWrap`, but I need to update it to check for `textWrapMode()` instead of calling a deprecated method. I'll do this in a follow up.

darinadler commented on Source/WebCore/layout/formattingContexts/inline/InlineQuirks.cpp:
```diff
@@ -44,7 +44,7 @@ bool InlineQuirks::trailingNonBreakingSpaceNeedsAdjustment(bool isInIntrinsicWid
         return false;
     auto& rootStyle = formattingContext().root().style();
-    return rootStyle.nbspMode() == NBSPMode::Space && rootStyle.textWrap() != TextWrap::NoWrap && rootStyle.whiteSpaceCollapse() != WhiteSpaceCollapse::BreakSpaces;
+    return rootStyle.nbspMode() == NBSPMode::Space && rootStyle.textWrapMode() != TextWrapMode::NoWrap && rootStyle.whiteSpaceCollapse() != WhiteSpaceCollapse::BreakSpaces;
```
I really wish we had a boolean for `text-wrap-mode` instead of an enumeration for call sites like this.

nt1m replied:
Same here

darinadler commented on Source/WebCore/layout/formattingContexts/inline/InlineTextItem.cpp:
```diff
@@ -89,7 +89,7 @@ bool InlineTextItem::isZeroWidthSpaceSeparator() const
 bool InlineTextItem::isQuirkNonBreakingSpace() const
 {
-    if (style().nbspMode() != NBSPMode::Space || style().textWrap() == TextWrap::NoWrap || style().whiteSpaceCollapse() == WhiteSpaceCollapse::BreakSpaces)
+    if (style().nbspMode() != NBSPMode::Space || style().textWrapMode() == TextWrapMode::NoWrap || style().whiteSpaceCollapse() == WhiteSpaceCollapse::BreakSpaces)
```
I really wish we had a boolean for `text-wrap-mode` instead of an enumeration for call sites like this.

nt1m replied:
Same here

darinadler commented on Source/WebCore/layout/formattingContexts/inline/TextOnlySimpleLineBuilder.cpp:
```diff
@@ -472,7 +472,7 @@ bool TextOnlySimpleLineBuilder::isEligibleForSimplifiedTextOnlyInlineLayout(cons
     if (rootStyle.hyphenationLimitLines() != RenderStyle::initialHyphenationLimitLines())
         return false;
-    if (rootStyle.textWrap() == TextWrap::Balance)
+    if (rootStyle.textWrapMode() == TextWrapMode::Wrap && rootStyle.textWrapStyle() == TextWrapStyle::Balance)
```
I really wish we had a boolean for `text-wrap-mode` instead of an enumeration for call sites like this.

nt1m replied:
Same here

darinadler commented on Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
```diff
@@ -383,13 +383,13 @@ bool TextUtil::shouldPreserveNewline(const Box& layoutBox)
 {
     // https://www.w3.org/TR/css-text-4/#text-wrap
-    return style.textWrap() != TextWrap::NoWrap;
+    return style.textWrapMode() != TextWrapMode::NoWrap;
```
I really wish we had a boolean for `text-wrap-mode` instead of an enumeration for call sites like this.

And then we could get rid of this function.

nt1m replied:
Same here, we should just use `style.autoWrap()` once it switches to use `text-wrap-mode`

darinadler commented on Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
```diff
@@ -383,13 +383,13 @@ bool TextUtil::shouldPreserveNewline(const Box& layoutBox)
 {
     // https://www.w3.org/TR/css-text-4/#white-space-phase-2
-    return style.whiteSpaceCollapse() == WhiteSpaceCollapse::Preserve && style.textWrap() != TextWrap::NoWrap;
+    return style.whiteSpaceCollapse() == WhiteSpaceCollapse::Preserve && style.textWrapMode() != TextWrapMode::NoWrap;
```
I really wish we had a boolean for `text-wrap-mode` instead of an enumeration for call sites like this.

nt1m replied:
ditto

darinadler commented on Source/WebCore/layout/integration/LayoutIntegrationCoverage.cpp:
```diff
@@ -195,7 +195,9 @@ bool shouldInvalidateLineLayoutPathAfterChangeFor(const RenderBlockFlow& rootBlo
     }
-    if (rootBlockContainer.style().direction() == TextDirection::RTL || rootBlockContainer.style().textWrap() == TextWrap::Balance)
+
+    bool shouldBalance = rootBlockContainer.style().textWrapMode() == TextWrapMode::Wrap && rootBlockContainer.style().textWrapStyle() == TextWrapStyle::Balance;
```
I really wish we had a boolean for `text-wrap-mode` instead of an enumeration for call sites like this.

nt1m replied:
ditto

darinadler commented on Source/WebCore/rendering/RenderBlock.cpp:
```diff
@@ -2268,7 +2268,7 @@ void RenderBlock::computeBlockPreferredLogicalWidths(LayoutUnit& minLogicalWidth
 
     const RenderStyle& styleToUse = style();
-    auto nowrap = styleToUse.textWrap() == TextWrap::NoWrap && styleToUse.whiteSpaceCollapse() == WhiteSpaceCollapse::Collapse;
+    auto nowrap = styleToUse.textWrapMode() == TextWrapMode::NoWrap && styleToUse.whiteSpaceCollapse() == WhiteSpaceCollapse::Collapse;
```
I really wish we had a boolean for `text-wrap-mode` instead of an enumeration for call sites like this.

nt1m replied:
ditto

darinadler commented on Source/WebCore/rendering/RenderListItem.cpp:
```diff
@@ -82,7 +82,7 @@ RenderStyle RenderListItem::computeMarkerStyle() const
     markerStyle.setUnicodeBidi(UnicodeBidi::Isolate);
     markerStyle.setWhiteSpaceCollapse(WhiteSpaceCollapse::Preserve);
-    markerStyle.setTextWrap(TextWrap::NoWrap);
+    markerStyle.setTextWrapMode(TextWrapMode::NoWrap);
```
Just confirming that we do not need to also set `text-wrap-style` here?

nt1m replied:
Same here, we don't need to balance the marker

darinadler commented on Source/WebCore/rendering/RenderText.cpp:
```diff
@@ -1336,7 +1336,7 @@ void RenderText::computePreferredLogicalWidths(float leadWidth, WeakHashSet<cons
         m_minWidth = m_maxWidth;
 
-    if (style.whiteSpaceCollapse() == WhiteSpaceCollapse::Preserve && style.textWrap() == TextWrap::NoWrap) {
+    if (style.whiteSpaceCollapse() == WhiteSpaceCollapse::Preserve && style.textWrapMode() == TextWrapMode::NoWrap) {
```
I really wish we had a boolean for `text-wrap-mode` instead of an enumeration for call sites like this.

nt1m replied:
ditto

darinadler commented on Source/WebCore/rendering/line/BreakingContext.h:
```diff
@@ -178,8 +178,8 @@ class BreakingContext {
 
-    TextWrap m_currentTextWrap { TextWrap::Wrap };
-    TextWrap m_lastObjectTextWrap { TextWrap::Wrap };
+    TextWrapMode m_currentTextWrap { TextWrapMode::Wrap };
+    TextWrapMode m_lastObjectTextWrap { TextWrapMode::Wrap };
```
Maybe these could just be booleans instead of using the enumeration?

nt1m replied:
It is possible, but I kind of like the explicitness of the enum values. Our style guide encourages this especially when passing arguments

darinadler replied:
Yes but we don't pass these as arguments. It's OK if you want to leave this but the enumeration suggestion for the style guide is to avoid mysterious `false` and `true` arguments and constants, which is not a thing here. I pushed for that style guideline and I think it's sometimes misunderstood.

darinadler commented on Source/WebCore/rendering/style/RenderStyleConstants.cpp:
```diff
@@ -1171,14 +1171,22 @@ TextStream& operator<<(TextStream& ts, TextUnderlinePosition underlinePosition)
+    switch (style) {
+    case TextWrapStyle::Auto: ts << "auto"; break;
+    case TextWrapStyle::Balance: ts << "balance"; break;
+    case TextWrapStyle::Stable: ts << "stable"; break;
```
Maybe do this alphabetically sorted, with `pretty` before `stable`?

nt1m replied:
Will fix.

darinadler commented on Source/WebCore/rendering/style/TextSizeAdjustment.cpp:
```diff
@@ -83,7 +83,7 @@ auto AutosizeStatus::computeStatus(const RenderStyle& style) -> AutosizeStatus
             return false;
 
-        if (style.whiteSpaceCollapse() == WhiteSpaceCollapse::Collapse && style.textWrap() == TextWrap::NoWrap)
+        if (style.whiteSpaceCollapse() == WhiteSpaceCollapse::Collapse && style.textWrapMode() == TextWrapMode::NoWrap)
```
I really wish we had a boolean for `text-wrap-mode` instead of an enumeration for call sites like this.

nt1m replied:
Same here

darinadler reviewed (approved):
Beautifully done.

nt1m committed:
[css-text-4] Update `text-wrap` properties to latest spec
https://bugs.webkit.org/show_bug.cgi?id=263432
rdar://117248327

Reviewed by Darin Adler.

There is now a new arrangement for the longhands that this commit implements:
- white-space is a shorthand of white-space-collapse & text-wrap-mode
- text-wrap is a shorthand of text-wrap-mode & text-wrap-style

The previous text-wrap property values are split between its longhands:
- text-wrap-style takes the values auto / balance / pretty / stable. This property is disabled behind a preference.
- text-wrap-mode takes the values wrap / nowrap.

This commit also enables by default the longhands for white-space now that layout support for all the different combinations is implemented.
It also removes the related workarounds.

WPTs are updated to use shortest serialization.

Specification: https://drafts.csswg.org/css-text-4/

* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/text-wrap-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/text-wrap-valid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/white-space-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/white-space-shorthand-text-wrap-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/white-space-shorthand-text-wrap.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/white-space-shorthand.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/white-space-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/white-space.html:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::toCSSValueID):
(WebCore::fromCSSValueID):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::textWrapShorthandValue const):
(WebCore::ComputedStyleExtractor::whiteSpaceShorthandValue const):
(WebCore::ComputedStyleExtractor::propertyValue const):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/css/ComputedStyleExtractor.h:
* Source/WebCore/css/ShorthandSerializer.cpp:
(WebCore::ShorthandSerializer::serialize):
(WebCore::ShorthandSerializer::serializeTextWrap const):
(WebCore::ShorthandSerializer::serializeWhiteSpace const):
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::CSSParserContext::CSSParserContext):
(WebCore::add):
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::initialValueForLonghand):
(WebCore::CSSPropertyParser::consumeTextWrapShorthand):
(WebCore::CSSPropertyParser::consumeWhiteSpaceShorthand):
(WebCore::CSSPropertyParser::parseShorthand):
* Source/WebCore/css/parser/CSSPropertyParser.h:
* Source/WebCore/editing/EditingStyle.cpp:
(WebCore::EditingStyle::conflictsWithInlineStyleOfElement const):
(WebCore::EditingStyle::removeStyleFromRulesAndContext):
(WebCore::StyleChange::StyleChange):
* Source/WebCore/editing/TextManipulationController.cpp:
(WebCore::isEnclosingItemBoundaryElement):
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::createInnerTextStyle):
* Source/WebCore/html/HTMLPreElement.cpp:
(WebCore::HTMLPreElement::collectPresentationalHintsForAttribute):
* Source/WebCore/html/HTMLTableCellElement.cpp:
(WebCore::HTMLTableCellElement::collectPresentationalHintsForAttribute):
* Source/WebCore/html/HTMLTextAreaElement.cpp:
(WebCore::HTMLTextAreaElement::collectPresentationalHintsForAttribute):
* Source/WebCore/html/track/TextTrackCueGeneric.cpp:
(WebCore::TextTrackCueGenericBoxElement::applyCSSProperties):
* Source/WebCore/html/track/VTTCue.cpp:
(WebCore::VTTCueBox::applyCSSProperties):
* Source/WebCore/layout/formattingContexts/inline/InlineFormattingContext.cpp:
(WebCore::Layout::InlineFormattingContext::layout):
* Source/WebCore/layout/formattingContexts/inline/InlineQuirks.cpp:
(WebCore::Layout::InlineQuirks::trailingNonBreakingSpaceNeedsAdjustment const):
* Source/WebCore/layout/formattingContexts/inline/InlineTextItem.cpp:
(WebCore::Layout::InlineTextItem::isQuirkNonBreakingSpace const):
* Source/WebCore/layout/formattingContexts/inline/TextOnlySimpleLineBuilder.cpp:
(WebCore::Layout::TextOnlySimpleLineBuilder::isEligibleForSimplifiedTextOnlyInlineLayout):
* Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
(WebCore::Layout::TextUtil::isWrappingAllowed):
(WebCore::Layout::TextUtil::shouldTrailingWhitespaceHang):
* Source/WebCore/layout/integration/LayoutIntegrationCoverage.cpp:
(WebCore::LayoutIntegration::shouldInvalidateLineLayoutPathAfterChangeFor):
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::computeBlockPreferredLogicalWidths const):
* Source/WebCore/rendering/RenderListItem.cpp:
(WebCore::RenderListItem::computeMarkerStyle const):
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::computePreferredLogicalWidths):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::adjustStyle):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::adjustMenuListStyle const):
* Source/WebCore/rendering/line/BreakingContext.h:
(WebCore::BreakingContext::clearLineBreakIfFitsOnLine):
(WebCore::BreakingContext::initializeForCurrentObject):
(WebCore::BreakingContext::handleReplaced):
(WebCore::BreakingContext::handleText):
(WebCore::BreakingContext::trailingSpacesHang):
(WebCore::BreakingContext::canBreakAtThisPosition):
(WebCore::BreakingContext::handleEndOfLine):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::RenderStyle):
(WebCore::RenderStyle::changeRequiresLayout const):
(WebCore::RenderStyle::conservativelyCollectChangedAnimatableProperties const):
(WebCore::RenderStyle::whiteSpace const):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::textWrapMode const):
(WebCore::RenderStyle::textWrapStyle const):
(WebCore::RenderStyle::setTextWrapMode):
(WebCore::RenderStyle::setTextWrapStyle):
(WebCore::RenderStyle::textWrap const): Deleted.
(WebCore::RenderStyle::setTextWrap): Deleted.
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator<<):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialTextWrapMode):
(WebCore::RenderStyle::initialTextWrapStyle):
(WebCore::RenderStyle::initialTextWrap): Deleted.
* Source/WebCore/rendering/style/TextSizeAdjustment.cpp:
(WebCore::AutosizeStatus::computeStatus):
* Source/WebCore/style/PropertyAllowlist.cpp:
(WebCore::Style::isValidMarkerStyleProperty):
(WebCore::Style::isValidCueStyleProperty):
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):
* Source/WebCore/svg/SVGTextContentElement.cpp:
(WebCore::SVGTextContentElement::collectPresentationalHintsForAttribute):

Canonical link: https://commits.webkit.org/269613@main
